includes/formatting.php

```diff
@@ -78,30 +78,83 @@ function sideload_images( $post_id ) {
 	require_once ABSPATH . 'wp-admin/includes/file.php';
 	require_once ABSPATH . 'wp-admin/includes/image.php';
 
-	$content = $post->post_content;
-	$pattern = '/["\'](https?:\/\/images.airstory.co\/[^"\']+)/i';
+	/*
+	 * Use DOMDocument to find all images in the post content.
+	 *
+	 * To avoid DOMDocument::saveHTML() from destroying the inner contents, we'll temporarily inject
+	 * a generic <div>.
+	 */
+	$use_internal = libxml_use_internal_errors( true );
+	$body         = new \DOMDocument;
+	$body->loadHTML( '<div>' . $post->post_content . '</div>', LIBXML_HTML_NODEFDTD | LIBXML_HTML_NOIMPLIED );
+	$images       = $body->getElementsByTagName( 'img' );
+	$pattern      = '/^https?:\/\/images.airstory.co\//i';
+	$replaced     = array();
+	$replacements = 0;
 
-	if ( ! preg_match_all( $pattern, $content, $matches ) ) {
-		return 0;
-	}
+	// Ensure media that gets sideloaded has the post_author set to the current user.
+	add_filter( 'wp_insert_attachment_data', __NAMESPACE__ . '\set_attachment_author' );
 
-	foreach ( array_unique( $matches['1'] ) as $remote ) {
-		$local = media_sideload_image( esc_url_raw( $remote ), $post_id, null, 'src' );
+	foreach ( $images as $image ) {
+		$src = $image->getAttribute( 'src' );
 
-		if ( is_wp_error( $local ) ) {
+		// Skip this image if it isn't Airstory-hosted media.
+		if ( ! preg_match( $pattern, $src ) ) {
 			continue;
 		}
 
-		$content = str_replace( $remote, $local, $content );
+		// Ensure we only sideload each piece of media once.
+		if ( isset( $replaced[ $src ] ) ) {
+			$local = $replaced[ $src ];
+
+		} else {
+			// Sideload the media.
+			$local = media_sideload_image( esc_url_raw( $src ), $post_id, null, 'src' );
```

Comment:
if you use `media_handle_sideload` instead of `media_sideload_image`, you'll have the attachment ID which will avoid the need to use the `get_attachment_id_by_url` function.

the setup is a bit different, but here's a working example: https://github.com/norcross/external-image-upload/blob/master/lib/helper.php#L89